transaction/transactor.go

```diff
@@ -15,57 +15,120 @@
+	ForwardedHeaders http.Header
+
+	//Body represents the full data off the XMiDT http.Response body
+	Body []byte
+}
+
+// Tr1d1umTransactor performs a typical HTTP request but
+// enforces some logic onto the HTTP transaction such as
+// context-based timeout and header filtering
+// this is a common utility for the stat and config tr1d1um services
+type Tr1d1umTransactor interface {
+	Transact(*http.Request) (*XmidtResponse, error)
+}
+
+// Tr1d1umTransactorOptions include parameters needed to configure the transactor
+type Tr1d1umTransactorOptions struct {
+	//RequestTimeout is the deadline duration for the HTTP transaction to be completed
+	RequestTimeout time.Duration
+
+	//Do is the core responsible to perform the actual HTTP request
+	Do func(*http.Request) (*http.Response, error)
+}
+
+type tr1d1umTransactor struct {
+	RequestTimeout time.Duration
+	Do             func(*http.Request) (*http.Response, error)
+}
+
+type TransactionRequest struct {
 	Address string `json:"address,omitempty"`
 	Path    string `json:"path,omitempty"`
 	Query   string `json:"query,omitempty"`
 	Method  string `json:"method,omitempty"`
 }
 
-func (re *transactionRequest) MarshalJSON() ([]byte, error) {
-	return json.Marshal(re)
-}
-
 type transactionResponse struct {
 	Code    int         `json:"code,omitempty"`
 	Headers interface{} `json:"headers,omitempty"`
 }
 
-type GetLoggerFunc func(context.Context) kitlog.Logger
+func (re *TransactionRequest) MarshalJSON() ([]byte, error) {
+	return json.Marshal(re)
+}
 
 func (rs *transactionResponse) MarshalJSON() ([]byte, error) {
 	return json.Marshal(rs)
 }
 
-// HeaderWPATID is the header key for the WebPA transaction UUID
-const HeaderWPATID = "X-WebPA-Transaction-Id"
+func NewTr1d1umTransactor(o *Tr1d1umTransactorOptions) Tr1d1umTransactor {
+	return &tr1d1umTransactor{
+		Do:             o.Do,
+		RequestTimeout: o.RequestTimeout,
+	}
+}
+
+func (t *tr1d1umTransactor) Transact(req *http.Request) (result *XmidtResponse, err error) {
+	ctx, cancel := context.WithTimeout(req.Context(), t.RequestTimeout)
+	defer cancel()
 
-// TransactionLogging is used by the different Tr1d1um services to
+	var resp *http.Response
+	if resp, err = t.Do(req.WithContext(ctx)); err == nil {
+		result = &XmidtResponse{
+			ForwardedHeaders: make(http.Header),
+			Body:             []byte{},
+		}
+
+		ForwardHeadersByPrefix("X", resp.Header, result.ForwardedHeaders)
+		result.Code = resp.StatusCode
+
+		defer resp.Body.Close()
+
+		result.Body, err = ioutil.ReadAll(resp.Body)
+		return
+	}
+
+	//Timeout, network errors, etc.
+	err = NewCodedError(err, http.StatusServiceUnavailable)
+	return
+}
+
+// Logging is used by the different Tr1d1um services to
 // keep track of incoming requests and their corresponding responses
-func TransactionLogging(reducedLoggingResponseCodes []int, logger kitlog.Logger) kithttp.ServerFinalizerFunc {
+func Logging(reducedLoggingResponseCodes []int, logger kitlog.Logger) kithttp.ServerFinalizerFunc {
```

Comment:
Logging is kind of weird, since normally our function verbs are more direct. 

```suggestion
func Log(reducedLoggingResponseCodes []int, logger kitlog.Logger) kithttp.ServerFinalizerFunc {
```

I also think the logger parameter should be first, but not sure that this pr is a good place to do that.

Reply:
I created a ticket for moving the parameter. I've applied the syntax to this pr!
https://github.com/xmidt-org/tr1d1um/issues/236